frame/contracts/src/storage/meter.rs

```diff
@@ -447,21 +463,30 @@ where
 		// We need to make sure that the contract's account exists.
 		T::Currency::transfer(origin, contract, ed, Preservation::Preserve)?;
 
+		System::<T>::inc_consumers(&contract)?;
+
 		// Normally, deposit charges are deferred to be able to coalesce them with refunds.
 		// However, we need to charge immediately so that the account is created before
 		// charges possibly below the ed are collected and fail.
-		E::charge(origin, contract, &deposit.saturating_sub(&Deposit::Charge(ed)), false)?;
+		E::charge(
+			origin,
+			contract,
+			&deposit.saturating_sub(&Deposit::Charge(ed)),
+			&ContractStatus::Alive,
+		)?;
 
 		Ok(deposit)
 	}
 
 	/// Call to tell the meter that the currently executing contract was executed.
 	///
 	/// This will manipulate the meter so that all storage deposit accumulated in
-	/// `contract_info` will be refunded to the `origin` of the meter.
-	pub fn terminate(&mut self, info: &ContractInfo<T>) {
+	/// `contract_info` will be refunded to the `origin` of the meter. And the free
+	/// (`reducible_balance`) will be sent to the `beneficiary`.
+	pub fn terminate(&mut self, info: &ContractInfo<T>, beneficiary: &T::AccountId) {
```

Comment:
There is no reason to pass by reference if we clone anyways. Just take by value and let the caller clone of appropriate.

```suggestion
	pub fn terminate(&mut self, info: &ContractInfo<T>, beneficiary: T::AccountId) {
```